pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go

```diff
@@ -645,3 +700,137 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14, 1),
+				makeCheckpointEvent([]byte("k"), []byte("m"), 12, 0),
+			},
+			expectedFrontierEntries: 4,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(0)},
+		},
+		"non-overlapping spans with close ts without quantization": {
+			withFrontierQuantization: false,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("z")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12, 3),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13, 2),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14, 1),
+				makeCheckpointEvent([]byte("k"), []byte("m"), 12, 0),
+			},
+			expectedFrontierEntries: 6,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(0)},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			frontier, err := span.MakeFrontier(tc.spans...)
+			require.NoError(t, err)
+			const quantize = time.Duration(10)
+			for _, e := range tc.events {
+				if tc.withFrontierQuantization {
+					e.Checkpoint.ResolvedTS = quantizeTS(e.Checkpoint.ResolvedTS, quantize)
+				}
+				_, err := frontier.Forward(e.Checkpoint.Span, e.Checkpoint.ResolvedTS)
+				require.NoError(t, err)
+			}
+			frontier.Entries(func(sp roachpb.Span, ts hlc.Timestamp) span.OpResult {
+				t.Logf("span: %v, ts: %v\n", sp, ts)
+				return false
+			})
+			require.Equal(t, tc.expectedFrontierEntries, frontier.Len())
+			require.Equal(t, tc.expectedFrontier, frontier.Frontier())
+		})
+	}
+}
+
+// TestFrontierQuantizationRand makes two frontiers with the same set of spans
+// and one with quantized ts and the other without. The test makes sure that the
+// highwater tracked by the frontier with quantized ts is <= the other, and the
+// number of spans being tracked is >= the other.
+func TestFrontierQuantizationRand(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	const numOfSpans = 100
+	events := make([]*kvpb.RangeFeedEvent, numOfSpans)
+	rng, _ := randutil.NewTestRand()
+	for i, s := range kvserverrangefeed.GenerateRandomizedSpans(rng, numOfSpans) {
+		ts := kvserverrangefeed.GenerateRandomizedTs(rng, time.Minute.Nanoseconds())
+		events[i] = makeCheckpointEvent(s.Key, s.EndKey, int(ts.WallTime), int32(rng.Intn(100)))
+	}
+
+	const quantize = time.Duration(10)
+	quantizedFrontier, err := span.MakeFrontier([]roachpb.Span{{Key: keys.MinKey, EndKey: keys.MaxKey}}...)
+	require.NoError(t, err)
+	for _, e := range events {
+		quantizedTs := quantizeTS(e.Checkpoint.ResolvedTS, quantize)
+		_, err := quantizedFrontier.Forward(e.Checkpoint.Span, quantizedTs)
+		require.NoError(t, err)
+	}
+	quantizedEntries := quantizedFrontier.Len()
+	quantizedHW := quantizedFrontier.Frontier()
+
+	frontier, err := span.MakeFrontier([]roachpb.Span{{Key: keys.MinKey, EndKey: keys.MaxKey}}...)
+	require.NoError(t, err)
+	for _, e := range events {
+		_, err := frontier.Forward(e.Checkpoint.Span, e.Checkpoint.ResolvedTS)
```

Comment:
nit: create both frontiers at the same time and have a single loop where you forward the events to both frontiers

Reply:
Done.